lightning/src/ln/onion_utils.rs

```diff
@@ -384,15 +382,25 @@ pub(super) fn process_onion_failure<T: secp256k1::Signing, L: Deref>(secp_ctx: &
 						|| error_code == 21; // Special case error 21 as the Route object is bogus, TODO: Maybe fail the node if the CLTV was reasonable?
 
 						let mut network_update = None;
+						let mut failing_route_hop = None;
 
 						if error_code & NODE == NODE {
 							network_update = Some(NetworkUpdate::NodeFailure { node_id: route_hop.pubkey, is_permanent: error_code & PERM == PERM });
 						}
 						else if error_code & PERM == PERM {
-							network_update = if payment_failed { None } else { Some(NetworkUpdate::ChannelClosed {
-								short_channel_id: path[next_route_hop_ix - if next_route_hop_ix == path.len() { 1 } else { 0 }].short_channel_id,
-								is_permanent: true,
-							})};
+							network_update = if payment_failed { None } else {
+								let failing_route_hop = if is_from_final_node { route_hop } else { &path[route_hop_idx + 1] };
```

Comment:
Actually, given we're finally doing retry scoring, I think we can drop the special-case of `error_code == 21` implying `payment_failed`, at which point `payment_failed` implies `is_from_final_node` and we can drop this if.